File: coding/neetcode_blind_75/reverse_linked_list.cpp
// Leetcode problem link - https://leetcode.com/problems/reverse-linked-list

// Solutions -

// 1. Three pointers
class Solution {
public:
    ListNode* reverseList(ListNode* head) {
        struct ListNode *current = head;
        struct ListNode *prev = NULL, *next = NULL;

        while (current != NULL) {
            next = current->next;
            current->next = prev;
            prev = current;
            current = next;
        }
        head = prev;
        
        return head;
    }
};
// Time Complexity - O(N)
// Space Complexity - O(N)
// Auxiliary space - O(1)

File: coding/striver_sheet/linked_list_1/reverse_linked_list.py
# Problem - https://leetcode.com/problems/reverse-linked-list

# Solution
# Definition for singly-linked list.
# class ListNode:
#     def __init__(self, val=0, next=None):
#         self.val = val
#         self.next = next

class Solution:
    def reverseList(self, head: Optional[ListNode]) -> Optional[ListNode]:
        current = head
        previous = next_node = None

        while current:
            next_node = current.next
            current.next = previous
            previous = current
            current = next_node

        head = previous

        return head
# Time complexity - O(n)
# Space comlexity - O(n)
# Auxiliary space - O(n)